doc/sphinx/proof-engine/tactics.rst

```diff
@@ -650,180 +668,381 @@ Applying theorems
+     the behavior depends on the form of the target. If the target is of the form
+     :n:`(fun x__1 … x__n => Q) u__1 … u__n` and the :n:`t__i` and :n:`u__i` unify,
+     then :g:`P` is instantiated into :n:`(fun x__1 … x__n => Q)`. Otherwise,
+     :tacn:`apply`
+     tries to define :g:`P` by abstracting over :n:`t__1 … t__n` in the target.
+     You can use :tacn:`pattern` to transform the target so that it
+     gets the form :n:`(fun x__1 … x__n => Q) u__1 … u__n`.  See the example
+     :ref:`here <example_apply_pattern>`.
+
+   :n:`@in_hyp_as` (the hypothesis case)
+     Proceeding from *right to left*, find the first premise of the type of
+     :n:`@one_term` that matches the specified hypothesis.  If a match
+     is found, the hypothesis is replaced with the conclusion of the type of
+     :n:`@one_term` (substituting for the unified variables)
+     and the tactic creates a new subgoal for each unmatched premise.
+     See the example :ref:`here <apply_forward>`.
+
+     If specified, :n:`as @simple_intropattern` is applied to the conclusion
+     of the type of :n:`@one_term`. In this case, the selected hypothesis
+     is left unchanged if its name is not reused.
+
+     If the type of :n:`@one_term` is an inductive type with a single constructor,
+     then each premise in the constructor is recursively matched to the conclusion
+     of the hypothesis and the first match is used.  See example :ref:`here <apply_with_iff>`.
+
+     For the hypothesis case, matching is done only with first-order unification.
+
+   :n:`with @bindings` (in :n:`@one_term_with_bindings`)
+     Gives explicit instantiations for variables used in the type of :n:`@one_term`.
+     There are 3 cases:
+
+     - Bindings for variables can be provided in a list of :n:`@one_term`\s
+       in the left-to-right order of their first appearance in the type of
+       :n:`@one_term`.  For the goal case (:ref:`example <apply_with_binding_goal>`),
+       the list should give bindings only for variables that aren't bound by
+       unification.  However, in the hypothesis case
+       (:ref:`example <apply_with_binding_hyp>`),
+       the list must include bindings for *all* variables.
+
+     - Bindings for unbound variables can be given by name with the
+       :n:`(@ident := @term)` form.
+
+     - The form :n:`(@natural := @term)` binds additional variables by
+       unifying the Nth premise of the type of :n:`@one_term` with :n:`@term`.
+       (Use `1` for the first premise.)
+
+   .. exn:: Unable to unify @one_term with @one_term.
+
+      The :tacn:`apply` tactic failed to match the conclusion of :token:`one_term`.
+      You can help :tacn:`apply` by
       transforming your goal with the :tacn:`change` or :tacn:`pattern`
       tactics.
 
+   .. exn:: Unable to apply lemma of type "..." on hypothesis of type "...".
+
+      This happens if the conclusion of :token:`ident` does not match any of
+      the premises of the type of :token:`term`.
```

Comment:
This should probably be:
```suggestion
      the premises of the type of :token:`one_term`.
```

Maybe also clarify that this is "the hypothesis case".

Reply:
I think the wording of the message `on hypothesis of type "..."` makes it clear that it's the hypothesis case.  Eh?